### PR TITLE
use custom --path-separator in arguments to --exec

### PR DESCRIPTION
**Pull request by aswild**
When --path-separator is used, pass it on to commands run with
--exec(-batch) and --list-details.

Fixes: https://github.com/sharkdp/fd/issues/697

### CODE REVIEW

**Comment by aswild** [comment]
There's definitely some corner cases (especially on Windows) that need to be tested here, especially when dealing with "absolute" paths. [`std::path::Prefix`](https://doc.rust-lang.org/std/path/enum.Prefix.html) can be six different things on Windows and I'm not sure how nicely they'll play with the new logic here.

Working with `OsStr(ing)` in Rust is as painful as ever, so I cast back to a `Path` and use its iterator over the components to re-join them with the custom separator.

I'll do some testing in MSYS when I get a chance.

**Comment by sharkdp** [comment]
Awesome - thank you for your contribution. Happy to review this once the tests are finished.

**Comment by aswild** [comment]
Refactored, tested on Windows, and force-pushed. I learned a lot more about Windows paths today.

Relative paths (originally with either `/` or `\` separators) work as expected, as do drive-letter absolute paths. UNC paths (`\\server\share\path`) also work, but only when using backslashes in `\\server\share` portion (Rust itself doesn't recognize `//server/share` as a UNC path).

Windows support may not be perfect for all the obscure corner cases, but if users are doing something weird, they can post-process fd's output using a tool like sed. Relative and drive-letter paths are surely the common use case, and here they work as expected.

File: src/exec/mod.rs [hunk 1]
```diff
@@ -201,21 +207,33 @@ impl ArgumentTemplate {
...
+    pub fn generate(&self, path: impl AsRef<Path>, path_separator: Option<&str>) -> OsString {
         use self::Token::*;
+        let path = path.as_ref();
 
         match *self {
             ArgumentTemplate::Tokens(ref tokens) => {
                 let mut s = OsString::new();
                 for token in tokens {
+                    // note: the s.push() and replace_separator() calls can't be pulled out of this
+                    // match or else we get errors about references to temporary values
```
Comment (sharkdp) [inline src/exec/mod.rs:222]: True, but I think that's to be expected. The alternative would be to copy or move the strings, but I think this is clearer.

If it's up to me, we can remove this comment.

File: src/exec/mod.rs [hunk 2]
```diff
@@ -224,6 +242,61 @@ impl ArgumentTemplate {
...
+    /// interpreted as a Path and its components are iterated through and re-joined into a new
+    /// OsString.
+    fn replace_separator<'a>(path: &'a OsStr, path_separator: Option<&str>) -> Cow<'a, OsStr> {
+        // fast-path - no replacement necessary
+        if path_separator.is_none() {
+            return Cow::Borrowed(path);
+        }
+
+        let path_separator = path_separator.unwrap();
+        let mut out = OsString::with_capacity(path.len());
```
Comment (sharkdp) [inline src/exec/mod.rs:257]: will only be enough if the path separator is on character, right? But okay, that's a strange corner case where we can live additional allocations, I guess.

File: src/exec/mod.rs [hunk 3]
```diff
@@ -315,26 +393,62 @@ mod tests {
...
+                assert_eq!(arg.generate($input, Some("#")), OsString::from($expected));
+            };
+        }
+
+        check!("foo", "foo");
+        check!("foo/bar", "foo#bar");
+        check!("/foo/bar/baz", "#foo#bar#baz");
+
+        // windows-specific absolute path prefixes
+        if cfg!(windows) {
```
Comment (sharkdp) [inline src/exec/mod.rs:435]: maybe move these to a separate `#[cfg(windows)]` test?

**Comment by sharkdp** [comment]
Very nice!

Thank you for your effort with testing this on Windows as well!

File: src/exec/mod.rs (hunk 1 above)
Comment (aswild) [reply src/exec/mod.rs:222]: Yeah, I think that expected, but I fought with the borrow checker a bit here so I tend to document that. I can remove the comment though if you prefer.

File: src/exec/mod.rs (hunk 2 above)
Comment (aswild) [reply src/exec/mod.rs:257]: Yeah, I think the Vec behind OsString grows by doubling in size so there's probably only one or two extra allocations for multi-character separators. The only way to know exactly how big out needs to be is by iterating & counting the path components beforehand, so just letting OsString automatically resize seems like a reasonable thing to do.

File: src/exec/mod.rs (hunk 3 above)
Comment (aswild) [reply src/exec/mod.rs:435]: sure, I'll do that

**Comment by sharkdp** [comment]
Thank you for the updates.
